server/src/main/java/org/opensearch/index/ShardIndexingPressureMemoryManager.java

```diff
@@ -0,0 +1,543 @@
+            the current node utilization is greater than primary_parameter.node.soft_limit of total node limits.
+             */
+            if(((double)nodeReplicaBytes / this.shardIndexingPressureSettings.getNodeReplicaLimits()) < this.nodeSoftLimit)  {
+                boolean isShardLimitsIncreased =
+                    this.increaseShardReplicaLimits(tracker, shardIndexingPressureStore);
+                if(isShardLimitsIncreased == false) {
+                    tracker.rejection().getReplicaNodeLimitsBreachedRejections().incrementAndGet();
+                    totalNodeLimitsBreachedRejections.incrementAndGet();
+                }
+
+                return !isShardLimitsIncreased;
+            } else {
+                boolean shardLastSuccessfulRequestDurationLimitsBreached =
+                    this.evaluateLastSuccessfulRequestDurationLimitsBreached(
+                        tracker.timeStamp().getLastSuccessfulReplicaRequestTimestamp().get(),
+                        requestStartTime, tracker.outstandingRequest().getTotalOutstandingReplicaRequests().get());
+
+                boolean shardThroughputDegradationLimitsBreached =
+                    this.evaluateThroughputDegradationLimitsBreached(
+                        Double.longBitsToDouble(tracker.throughput().getReplicaThroughputMovingAverage().get()),
+                        tracker.memory().getTotalReplicaBytes().get(), tracker.latency().getReplicaTimeInMillis().get(),
+                        tracker.throughput().getReplicaThroughputMovingQueue().size(), replicaThroughputDegradationLimits);
+
+                if (shardLastSuccessfulRequestDurationLimitsBreached || shardThroughputDegradationLimitsBreached) {
+                    if(shardLastSuccessfulRequestDurationLimitsBreached) {
+                        tracker.rejection().getReplicaLastSuccessfulRequestLimitsBreachedRejections().incrementAndGet();
+                        totalLastSuccessfulRequestLimitsBreachedRejections.incrementAndGet();
+                    } else {
+                        tracker.rejection().getReplicaThroughputDegradationLimitsBreachedRejections().incrementAndGet();
+                        totalThroughputDegradationLimitsBreachedRejections.incrementAndGet();
+                    }
+
+                    return true;
+                } else {
+                    boolean isShardLimitsIncreased =
+                        this.increaseShardReplicaLimits(tracker, shardIndexingPressureStore);
+                    if(isShardLimitsIncreased == false) {
+                        tracker.rejection().getReplicaNodeLimitsBreachedRejections().incrementAndGet();
+                        totalNodeLimitsBreachedRejections.incrementAndGet();
+                    }
+
+                    return !isShardLimitsIncreased;
+                }
+            }
+        } else {
+            return false;
+        }
+    }
```

Comment:
Lets add this to meta as a fast follow up to break this monolith

Reply:
Yes, added in #478 as an item list.